### PR TITLE
Integrate E2B sandbox as an alternative to a Docker container

### PR DESCRIPTION
**Pull request by mlejva**
This is a draft PR for integrating [E2B](https://e2b.dev/) sandbox for OpenDevin's agents. The E2B sandbox is a secure micro VM made for running AI agents and AI generated code in the cloud.

I'll be updating the following todo list as I go:
- [x] Bring E2B sandbox to the parity in terms of functionality with the current `sandbox.py`
- [x] Common interface between Docker container and E2B sandbox
- [x] Update fileops to happen inside sandbox when using E2B
- [ ] ~Use E2B sandbox for running swe-bench evals end-to-end~ (will be done in separate PR)

### CODE REVIEW

**Comment by foragerr** [comment]
curious: Can e2b run inside docker?

**Comment by mlejva** [comment]
> curious: Can e2b run inside docker?

No. E2B sandbox is a VM so it sits a "lower level" than a container. What will be possible in the future is to run Docker inside E2B.

File: opendevin/controller/command_manager.py
Comment (rbren) [inline opendevin/controller/command_manager.py:18]: We should probably switch this based on an env var

File: opendevin/controller/command_manager.py
Comment (rbren) [inline opendevin/controller/command_manager.py:23]: we'll definitely need background support before merging this--otherwise the agents will get stuck.

Does e2b support long-running commands? E.g. can the agent run `node server.js` and then `curl localhost:3000`?

**Review by rbren — COMMENTED** [review]
Looking good!

We can probably formalize the sandbox interface with an abstract class--doesn't need to be in this PR though

File: opendevin/controller/command_manager.py
Comment (mlejva) [reply opendevin/controller/command_manager.py:23]: Yes, that's possible. Each sandbox has it's own URL and we automatically allow you to connect to a port when it opens. We support HTTP request and Websockets at the moment. This way you can connect for example to a running Nextjs app that you start with `next dev` and see how OpenDevin makes changes in real-time.

**Comment by mlejva** [comment]
@rbren a short update, still work in progress.

I updated the PR so the E2B sandbox matches the functionality of the current Docker sandbox. Specifically the background commands are now supported.

Next step is to:
- [] clean up and improve types so we can build a shared interface between E2B sandbox and Docker sandbox
- [] enable filesystem operations. Currently, the filesystem operations with Docker sandbox are happening through a mounted volume because the container is running locally. WIth E2B sandbox
- [] add config option to allow switching between E2B sandbox and Docker sandbox

I think I should be able to finish the whole PR this Sunday

**Comment by mlejva** [comment]
@rbren similar to what @foragerr did in #874 I think we should also build a common interface for process (what's known as command at the moment). It will be something fairly simple like


```python
from abc import ABC, abstractmethod

class Process(ABC):
    @abstractmethod
    def kill(self):
        pass

    @abstractmethod
    def read_logs(self) -> str:
        pass
    
    @property
    @abstractmethod
    def id(self) -> int:
        pass

    @property
    @abstractmethod
    def command(self) -> str:
        pass

    @property
    @abstractmethod
    def exit_code(self) -> int | None: # The reason it can be none is because the process can still be running
       pass 
```

What do you think?

**Comment by rbren** [comment]
@mlejva that seems reasonable to me!

**Comment by mlejva** [comment]
> curious: Can e2b run inside docker?

@foragerr I realized I might have misunderstood your question. I thought you're talking about self-hosting. With self-hosting what I wrote stands - E2B sandbox is a VM so it sits on a "lower level" compared to a container. What will be possible in the future is to run Docker inside E2B.

But you can definitely use E2B SDK for launching sandboxes in our cloud from Docker and Dockerize the whole app. 

I just wanted to clear that up

File: opendevin/sandbox/e2b/sandbox.py
Comment (neubig) [inline opendevin/sandbox/e2b/sandbox.py:25]: Just to confirm, since I haven't run this, is a good error thrown if `E2B_API_KEY` is not set? I feel that we should probably check that the key is set and throw a comprehensible error if not.

File: opendevin/sandbox/process.py
Comment (neubig) [inline opendevin/sandbox/process.py:22]:
```suggestion
```

Delete commented code, unless it's commented with the intention to uncomment it eventually, in which case we can add a TODO.

**Review by neubig — COMMENTED** [review]
Looks great, thanks a bunch for the contribution! I just have a few small comments.

File: opendevin/sandbox/e2b/e2b.Dockerfile
Comment (mlejva) [reply opendevin/sandbox/e2b/e2b.Dockerfile:1]: Typically our users have `Dockerfile` for their whole app in the same directory as a Dockerfile for E2B sandbox. We wanted to make it easy to distinguish for them so we went with `e2b.Dockerfile`.  No problem at all with renaming it to `Dockerfile` though!

File: opendevin/sandbox/process.py
Comment (mlejva) [reply opendevin/sandbox/process.py:22]: Sorry, you're right, it shouldn't be there.

File: opendevin/sandbox/e2b/sandbox.py
Comment (mlejva) [reply opendevin/sandbox/e2b/sandbox.py:25]: We throw an exception if not API key is set - https://github.com/e2b-dev/E2B/blob/1221a57e7ec49d2e9fbc5ca1804d201bf5fa3673/packages/python-sdk/e2b/utils/api_key.py#L11

Would that be suffice?

**Comment by mlejva** [comment]
Thank you for the review @neubig! I resolved the comments and will fix the build errors today. I'm on the move so I'll get back to my laptop only tonight.

**Review by neubig — APPROVED** [review]
Thanks! LGTM once the errors are resolved.

File: opendevin/action/fileop.py
Comment (rbren) [inline opendevin/action/fileop.py:28]:
```suggestion
```

File: opendevin/action/fileop.py
Comment (rbren) [inline opendevin/action/fileop.py:50]:
```suggestion
```

File: opendevin/action/fileop.py
Comment (rbren) [inline opendevin/action/fileop.py:49]: This feels leaky to me. But passing `controller` into `run` is already kind of leaky. Curious if you have any ideas on how to improve this.

Maybe `sandbox` should handle all the `run` logic...

**Review by rbren — APPROVED** [review]
Excited to get this in!

File: opendevin/action/fileop.py
Comment (mlejva) [reply opendevin/action/fileop.py:49]: Yeah I agree this doesn't feel right. Happy to refactorize this

**Comment by mlejva** [comment]
Hey @rbren thank you so much for the code review. I merge the `main` branch. I'm struggling a bit with understanding why the build didn't pass though.

I built it locally with both Python 3.11 and 3.12 and it looks like it's passing.

Screenshot 2024-04-15 at 11 24 41 PM

Screenshot 2024-04-15 at 11 23 31 PM

 The errors in GitHub actions seem to only show error code and nothing else though:
- https://github.com/OpenDevin/OpenDevin/actions/runs/8675380093/job/23793473005?pr=727
- https://github.com/OpenDevin/OpenDevin/actions/runs/8675380093/job/23793472708?pr=727
- https://github.com/OpenDevin/OpenDevin/actions/runs/8675380093/job/23793472924?pr=727

Screenshot 2024-04-15 at 11 25 20 PM


I'm not sure why it's not passing, I you have any tips, it would be super helpful!

**Comment by mlejva** [comment]
I also did a bit of refactorization in this PR. 

- `Sandbox` is now separated to [`docker`](https://github.com/mlejva/OpenDevin/tree/e2b-sandbox/opendevin/sandbox/docker) and [`e2b`](https://github.com/mlejva/OpenDevin/tree/e2b-sandbox/opendevin/sandbox/e2b)
- `BackgroundCommand` is now `Process` with separate `
- I added [readme to E2B sandbox](https://github.com/mlejva/OpenDevin/blob/e2b-sandbox/opendevin/sandbox/e2b/README.md) to get people started easily
- I pre-built a [custom E2B sandbox](https://github.com/mlejva/OpenDevin/tree/e2b-sandbox/containers/e2b-sandbox) for OpenDevin called `open-devin`.
    - Currently, it's only 0.5GB RAM and 2vCPU and people need to use their on SDK key but once we get OpenDevin to a cloud hosted webapp, we'll upgrade it to better machine, and sponsor all needed compute for OpenDevin's team API key

File: containers/e2b-sandbox/README.md
Comment (rbren) [inline containers/e2b-sandbox/README.md:14]: Can we `e2b template` from a remote image like `ubuntu` or `ghcr.io/opendevin/sandbox`? Or does the user have to build it from scratch?

Would be nice not to have to maintain a new Dockerfile, or to at least build it for them in our CI/CD pipeline

File: containers/e2b-sandbox/e2b.toml
Comment (rbren) [inline containers/e2b-sandbox/e2b.toml:1]: I don't see this file getting used anywhere?

**Review by rbren — APPROVED** [review]
This LGTM as-is! A couple minor notes though

File: opendevin/sandbox/e2b/README.md
Comment (rbren) [inline opendevin/sandbox/e2b/README.md:9]: We're mostly getting rid of `config.toml` now that everything is running inside docker. Maybe this should be passed into `docker run` with `-e`?

File: containers/e2b-sandbox/e2b.toml
Comment (mlejva) [reply containers/e2b-sandbox/e2b.toml:1]: It's used by the E2B CLI, it's something like Cloudflare Workers have their [`wrangler.toml`](https://developers.cloudflare.com/workers/wrangler/configuration/)

File: containers/e2b-sandbox/README.md
Comment (mlejva) [reply containers/e2b-sandbox/README.md:14]: You can share the same Dockerfile and use a remote image. We just have some restrictions on what images and Dockerfile commands we support. Generally, you should be good. I'll set it up to use the same Dockerfile image

Screenshot 2024-04-16 at 2 53 10 PM

File: opendevin/sandbox/e2b/README.md
Comment (mlejva) [reply opendevin/sandbox/e2b/README.md:9]: Yeah, that should work, I can update it. So `config.py` isn't used anymore and instead we should pass the env vars, right?

Comment (mlejva) [reply opendevin/sandbox/e2b/README.md]: I updated the readme to instruct users to set the E2B API key to the `E2B_API_KEY` env var. 

If there's nothing else you think should change, this PR should be good to go!

**Comment by rbren** [comment]
auto-merge enabled! 🚀 

Thanks for all the work here!

**Comment by rbren** [comment]
Sorry, merge conflicts 😭 

SWE-agent implemented start and end lines for read/write operations, which I suppose we'll need to support with e2b. Hopefully that's not too hard...

**Comment by mlejva** [comment]
> Sorry, merge conflicts 😭
> 
> SWE-agent implemented start and end lines for read/write operations, which I suppose we'll need to support with e2b. Hopefully that's not too hard...

Looking into it in 1-2 hours!

**Comment by mlejva** [comment]
Hey @rbren I finally got to resolving the conflicts. Sorry it took longer. 
Can you please check it when you have a moment and if it looks good to you merge it? 

Thank you!

**Pull request merged**
